### DISCUSSION TITLE
baresip for iOS

### DISCUSSION THREAD

**Discussion by stan-phillips** [baresip/baresip#3144] [Q&A]
I have been tasked with compiling baresip for use in an iOS project. I have been able to successfully compile the re library using an iOS toolchain file. I point to that when compiling baresip as it is a required dependency. I am seeing errors suggesting that some of the modules like vumeter and vidinfo are not compiled for the correct architecture. Has anyone here been able to successfully compile baresip for use in an iOS project and if so do you have any info on this?

**Comment by frogblueDFi** [comment]
Hi, maybe late but maybe helpful.

Yes I have been able to cross-compile all the necessary dependencies to use baresip as a library in an iOS application for video and audio SIP communication. Also for macOS and Android.
I cross compiled openssl, re, lib tiff, spendsp, x264, x265, ffmpeg and finally baresip.
When building baresip I explicitly add modules to compile: auconv, auresamp, g711, g722, menu, netroam, stun, turn, ice, vumeter, avcodec, avfilter, avformat, swscale and a bunch of custom and patched/reworked other modules.

So believe me, it is a painful job as most build scripts you can find will not work out of the box.
Guess you have to do it on your own.

Good luck.
Feel free to ask if you need more advice.

**Reply by juha-h** [reply]
Since this is an open source project, do you have somewhere the build scripts available?

**Reply by SnowMaster12** [reply]
Hi @frogblueDFi.
Out of curiosity, do you also experience lag with h264 video calls?
I tried with Linphone and the video stream reaches the Linphone client, but it lags a lot, even when lowering the bitrate significantly. 
i'm using baresip 4.3.0, with ffmpeg 6.1.4 ( i'm using videotoolbox for encoding )

**Reply by frogblueDFi** [reply]
Hi @SnowMaster12
No, I don't have any lags. Video performance is quite good. I still use baresip 3.24 but ffmpeg n7.1.
But this should not matter.
Even if I set „unfavourable“ values for bitrate, send_bitrate and burstbits I see a bad video, but no relevant latency.
Do you also use baresip as a lib compiled  for iOS?
Dietmar

**Reply by SnowMaster12** [reply]
@frogblueDFi Yes, I’m using baresip as a compiled library for iOS.
I also had to cross-compile FFmpeg to make it work on iOS, and it was a nightmare (same story with avcapture). Audio calls work perfectly, but video doesn’t, on Linphone the video arrives with noticeable delay, while other SIP clients report that my H.264 is “broken".

**Reply by frogblueDFi** [reply]
@SnowMaster12
Do you use Videotoolbox hardware acceleration successfully on iOS? Or h.264 software encoding only with x264 lib?

**Reply by SnowMaster12** [reply]
@frogblueDFi 
Yup I’m using VideoToolbox hardware acceleration on iOS ( only for encoding )
`avcodec: using H.264 encoder 'h264_videotoolbox' -- VideoToolbox H.264 Encoder`
`avcodec: using H.264 decoder 'h264' -- H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10`

**Comment by frogblueDFi** [comment]
Hi Juha-h!

Not available in any public repo.
But I could provide them, if someone would like to create a repo and maintain them.

Most of the work has been to find the correct parameters for cmake and configure.
They are based on baresip 3.8.1. Guess some other patches are required for newer versions.
And I am almost sure that they will not run on other machines out of the box.
But nevertheless a very good starting point!

Dietmar

**Reply by juha-h** [reply]
I have repo https://github.com/juha-h/libbaresip-android for Android libs.  Video libs are in video branch.  The scripts are using ffmpeg-kit, but that kit is not maintained anymore and the used ffmpeg version is an old one.  So I have been thinking to get rid of ffmpeg-kit dependency and build each lib separately based on latest versions.  When I have time, it would be interesting to compare to yours if you could put your scripts available somewhere.

**Reply by frogblueDFi** [reply]
Check if you have access to https://github.com/frogblueDFi/BuildBaresipLibs

**Reply by xdien** [reply]
> Check if you have access to https://github.com/frogblueDFi/BuildBaresipLibs

I am also facing some similar issues while building iOS, please give me access to that repo

**Reply by marain87** [reply]
@frogblueDFi @juha-h, I am also facing this issue ... Could you please share the build script with me too?

**Reply by xdien** [reply]
> Check if you have access to https://github.com/frogblueDFi/BuildBaresipLibs

Thank you! I got in.

**Reply by marain87** [reply]
> Check if you have access to https://github.com/frogblueDFi/BuildBaresipLibs

Thank you very much.

**Reply by MortenBuderich123** [reply]
@frogblueDFi  @juha-h , I have the same issue. Could you please give me access too?

**Reply by code4food-io** [reply]
@frogblueDFi , Could you please give me access too?

**Reply by code4food-io** [reply]
@frogblueDFi Thank you very much.

**Reply by fosple** [reply]
Could you also give me access as I'm also struggling @frogblueDFi

**Comment by juha-h** [comment]
frogblueDFi writes:

> Check if you have access to > https://github.com/frogblueDFi/BuildBaresipLibs

Thanks, access worked.

**Comment by juha-h** [comment]
ffmpeg version 6.0 seems to be the same as in ffmpeg-kit.  Latest ffmpeg
release is 7.1.

**Reply by frogblueDFi** [reply]
If the scripts are running on your system (I implement and cross compile almost everything on macOS), it is easy to give all the different versions a try. Different patches may be needed to compile and link.
I also plan to do a version jump in particular for baresip. I am way behind.

**Reply by frogblueDFi** [reply]
Did not yet try do all this on Windows, guess this another painful job!
